### PR TITLE
[mail] Fix set UTF-8 values to headers. Fix #7501

### LINKED ISSUE #7501: invalid german umlauts üöä Zend\Mail\Header\HeaderValue
How do I have to encode german umlauts like ö ä ü for Zend\Mail\Header\Subject? 

In my mail everything is UTF8 encoded. Did I missed something?

Zend\Mail\Header\HeaderValue::isValid() says they are invalid.
Zend\Mail\Header\HeaderValue::filter() strips all umlauts.

ZF Version: 2.4.1
Security Advisory: http://framework.zend.com/security/advisory/ZF2015-04

### PR DESCRIPTION
**Pull request by Maks3w**

### CODE REVIEW

File: library/Zend/Mail/Address.php [hunk 1]
```diff
@@ -24,15 +27,33 @@ class Address implements Address\AddressInterface
...
     public function __construct($email, $name = null)
     {
-        if (!is_string($email)) {
-            throw new Exception\InvalidArgumentException('Email must be a string');
+        $emailAddressValidator = new EmailAddressValidator(Hostname::ALLOW_LOCAL);
+        if (!is_string($email) || empty($email)) {
+            throw new Exception\InvalidArgumentException('Email must be a valid email address');
+        }
+
+        if (preg_match("/[\r|\n]/", $email)) {
```
Comment (weierophinney) [inline library/Zend/Mail/Address.php]: You can remove the `|` character as this is a character class.

File: library/Zend/Mail/Address.php [hunk 2]
```diff
@@ -24,15 +27,33 @@ class Address implements Address\AddressInterface
...
         }
-        if (null !== $name && !is_string($name)) {
-            throw new Exception\InvalidArgumentException('Name must be a string');
+
+        if (null !== $name) {
+            if (!is_string($name)) {
+                throw new Exception\InvalidArgumentException('Name must be a string');
+            }
+
+            if (preg_match("/[\r|\n]/", $name)) {
```
Comment (weierophinney) [inline library/Zend/Mail/Address.php]: Remove the `|` character.

File: library/Zend/Mail/Address.php (hunk 1 above)
Comment (Maks3w) [inline library/Zend/Mail/Address.php]: done

File: library/Zend/Mail/Header/GenericHeader.php [hunk 3]
```diff
@@ -163,6 +155,14 @@ public function setEncoding($encoding)
 
     public function getEncoding()
     {
+        if (!$this->encoding) {
+            if (Mime::isPrintable($this->fieldValue)) {
+                $this->setEncoding('ASCII');
+            } else {
+                $this->setEncoding('UTF-8');
+            }
```
Comment (weierophinney) [inline library/Zend/Mail/Header/GenericHeader.php]: Let's make this a ternary:

```php
$this->encoding = Mime::isPrintable($this->fieldValue) ? 'ASCII' : 'UTF-8';
```

File: library/Zend/Mail/Header/Sender.php [hunk 4]
```diff
@@ -93,6 +93,14 @@ public function setEncoding($encoding)
 
     public function getEncoding()
     {
+        if (!$this->encoding) {
+            if (Mime::isPrintable($this->getFieldValue(HeaderInterface::FORMAT_RAW))) {
+                $this->setEncoding('ASCII');
+            } else {
+                $this->setEncoding('UTF-8');
+            }
```
Comment (weierophinney) [inline library/Zend/Mail/Header/Sender.php]: Same comment as above; make this a ternary.

File: library/Zend/Mail/Header/Subject.php [hunk 5]
```diff
@@ -66,16 +69,29 @@ public function setEncoding($encoding)
 
     public function getEncoding()
     {
+        if (!$this->encoding) {
+            if (Mime::isPrintable($this->subject)) {
+                $this->setEncoding('ASCII');
+            } else {
+                $this->setEncoding('UTF-8');
+            }
+        }
```
Comment (weierophinney) [inline library/Zend/Mail/Header/Subject.php]: Same.

**Comment by CreativeNative** [comment]
Still get the same error.
Zend\Mail\Header\Exception\InvalidArgumentException
"Invalid Subject value detected"

.../vendor/zendframework/zendframework/library/Zend/Mail/Message.php:336 Zend\Mail\Header\Subject->setSubject ("Buchungsanfrage für Castello Rosso II" )

ZF 2.4.2dev

**Pull request merged**